

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>ccanalyser cli documentation &mdash; ccanalyser 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="ccanalyser tools" href="module_tools/tools.html" />
    <link rel="prev" title="Running the pipeline" href="running_the_pipeline.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ccanalyser
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installing.html">Installing ccanalyser</a></li>
<li class="toctree-l1"><a class="reference internal" href="running_the_pipeline.html">Running the pipeline</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">ccanalyser cli documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#ccanalyser">ccanalyser</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-fastq-deduplicate">fastq-deduplicate</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-fastq-deduplicate-identify">identify</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-fastq-deduplicate-parse">parse</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-fastq-deduplicate-remove">remove</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-fastq-digest">fastq-digest</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-fastq-split">fastq-split</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-genome-digest">genome-digest</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-interactions-bedgraph">interactions-bedgraph</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-interactions-count">interactions-count</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-interactions-store">interactions-store</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-interactions-store-bins">bins</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-interactions-store-fragments">fragments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-interactions-store-merge">merge</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-reporters-deduplicate">reporters-deduplicate</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-reporters-deduplicate-identify">identify</a></li>
<li class="toctree-l4"><a class="reference internal" href="#ccanalyser-reporters-deduplicate-remove">remove</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-reporters-identify">reporters-identify</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ccanalyser-slices-annotate">slices-annotate</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="module_tools/tools.html">ccanalyser tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="module_cli/modules.html">cli</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ccanalyser</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>ccanalyser cli documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/ccanalyser.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="ccanalyser-cli-documentation">
<h1>ccanalyser cli documentation<a class="headerlink" href="#ccanalyser-cli-documentation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="ccanalyser">
<h2>ccanalyser<a class="headerlink" href="#ccanalyser" title="Permalink to this headline">¶</a></h2>
<p class="rubric">Parameters</p>
<p>Base of all ccanalyser cli commands.
Type -h or –help after any subcommand for more information.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<div class="section" id="ccanalyser-fastq-deduplicate">
<h3>fastq-deduplicate<a class="headerlink" href="#ccanalyser-fastq-deduplicate" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Identifies PCR duplicate fragments from Fastq files</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-deduplicate <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<div class="section" id="ccanalyser-fastq-deduplicate-identify">
<h4>identify<a class="headerlink" href="#ccanalyser-fastq-deduplicate-identify" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Identifies fragments with duplicated sequences.</p>
<blockquote>
<div><p>Takes the input of the parse subcommand and deduplicates the hashed sequences
using a python dictionary. Duplicate read ids (hashed) are output  as a .json file.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">param input_files</dt>
<dd class="field-odd"><p>Paths to json files containing dictionaries with hashed read ids as the keys
and hashed sequences as the values.</p>
</dd>
<dt class="field-even">type input_files</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Duplicate read ids identified. Defaults to “duplicates.json”.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-deduplicate identify <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILES<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-identify-o">
<span id="cmdoption-ccanalyser-fastq-deduplicate-identify-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-identify-o" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Output file</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-identify-arg-INPUT_FILES">
<span id="cmdoption-ccanalyser-fastq-deduplicate-identify-arg-input-files"></span><code class="sig-name descname"><span class="pre">INPUT_FILES</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-identify-arg-INPUT_FILES" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument(s)</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-fastq-deduplicate-parse">
<h4>parse<a class="headerlink" href="#ccanalyser-fastq-deduplicate-parse" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Parses fastq file(s) into easy to deduplicate format.</p>
<p>Generates a hashed (using xxhash) dictionary in json format for deduplication using identify.</p>
<dl class="field-list simple">
<dt class="field-odd">param input_files</dt>
<dd class="field-odd"><p>One or more fastq files to process</p>
</dd>
<dt class="field-even">type input_files</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output for parsed read identifiers and sequences. Defaults to “out.json”.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param read_buffer</dt>
<dd class="field-odd"><p>Number of reads to process before outputting to file. Defaults to 1e5.</p>
</dd>
<dt class="field-even">type read_buffer</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-deduplicate parse <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILES<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-parse-o">
<span id="cmdoption-ccanalyser-fastq-deduplicate-parse-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-parse-o" title="Permalink to this definition">¶</a></dt>
<dd><p>File to store hashed sequence identifiers</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-parse-read_buffer">
<span id="cmdoption-ccanalyser-fastq-deduplicate-parse-read-buffer"></span><code class="sig-name descname"><span class="pre">--read_buffer</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_buffer&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-parse-read_buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of reads to process before writing to file</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-parse-arg-INPUT_FILES">
<span id="cmdoption-ccanalyser-fastq-deduplicate-parse-arg-input-files"></span><code class="sig-name descname"><span class="pre">INPUT_FILES</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-parse-arg-INPUT_FILES" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument(s)</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-fastq-deduplicate-remove">
<h4>remove<a class="headerlink" href="#ccanalyser-fastq-deduplicate-remove" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Removes fragments with duplicated sequences from fastq files.</p>
<p>Reads fastq files and removes all duplicate entries provided by the duplicated ids file.
Uses the output of the identify subcommand to identify and remove all sequence duplicates.</p>
<dl class="field-list simple">
<dt class="field-odd">param input_files</dt>
<dd class="field-odd"><p>Input fastq files (in the same order as used for the parse command).</p>
</dd>
<dt class="field-even">type input_files</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param duplicated_ids</dt>
<dd class="field-odd"><p>Duplicated read ids from identify command (hashed and in json format).</p>
</dd>
<dt class="field-even">type duplicated_ids</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param read_buffer</dt>
<dd class="field-odd"><p>Number of reads to process before writing to file. Defaults to 1e5.</p>
</dd>
<dt class="field-even">type read_buffer</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param output_prefix</dt>
<dd class="field-odd"><p>Deduplicated fastq output prefix. Defaults to “”.</p>
</dd>
<dt class="field-even">type output_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param gzip</dt>
<dd class="field-odd"><p>Determines if output is gzip compressed using pigz. Defaults to False.</p>
</dd>
<dt class="field-even">type gzip</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param compression_level</dt>
<dd class="field-odd"><p>Level of compression if required (1-9). Defaults to 5.</p>
</dd>
<dt class="field-even">type compression_level</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param sample_name</dt>
<dd class="field-odd"><p>Name of sample processed e.g. DOX-treated_1. Defaults to “”.</p>
</dd>
<dt class="field-even">type sample_name</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param stats_prefix</dt>
<dd class="field-odd"><p>Output prefix for statistics. Defaults to “”.</p>
</dd>
<dt class="field-even">type stats_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-deduplicate remove <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILES<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-o">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-output_prefix"></span><span id="cmdoption-ccanalyser-fastq-deduplicate-remove-output-prefix"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for deduplicated fastq file(s)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-d">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-duplicated_ids"></span><span id="cmdoption-ccanalyser-fastq-deduplicate-remove-duplicated-ids"></span><code class="sig-name descname"><span class="pre">-d</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--duplicated_ids</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;duplicated_ids&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to duplicate ids, identified by the identify subcommand</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-read_buffer">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-read-buffer"></span><code class="sig-name descname"><span class="pre">--read_buffer</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_buffer&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-read_buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of reads to process before writing to file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-gzip">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-no-gzip"></span><code class="sig-name descname"><span class="pre">--gzip</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--no-gzip</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-gzip" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines if files are gziped or not</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-compression_level">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-compression-level"></span><code class="sig-name descname"><span class="pre">--compression_level</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;compression_level&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-compression_level" title="Permalink to this definition">¶</a></dt>
<dd><p>Level of compression for output files</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-sample_name">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-sample-name"></span><code class="sig-name descname"><span class="pre">--sample_name</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;sample_name&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-sample_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of sample e.g. DOX_treated_1</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-stats_prefix">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-stats-prefix"></span><code class="sig-name descname"><span class="pre">--stats_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;stats_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-stats_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for stats file</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-deduplicate-remove-arg-INPUT_FILES">
<span id="cmdoption-ccanalyser-fastq-deduplicate-remove-arg-input-files"></span><code class="sig-name descname"><span class="pre">INPUT_FILES</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-deduplicate-remove-arg-INPUT_FILES" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument(s)</p>
</dd></dl>

</div>
</div>
<div class="section" id="ccanalyser-fastq-digest">
<h3>fastq-digest<a class="headerlink" href="#ccanalyser-fastq-digest" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Performs in silico digestion of one or a pair of fastq files.</p>
<dl class="field-list simple">
<dt class="field-odd">param input_fastq</dt>
<dd class="field-odd"><p>Input fastq files to process</p>
</dd>
<dt class="field-even">type input_fastq</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param restriction_enzyme</dt>
<dd class="field-odd"><p>Restriction enzyme name or site to use for digestion.</p>
</dd>
<dt class="field-even">type restriction_enzyme</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param mode</dt>
<dd class="field-odd"><p>Digest combined(flashed) or non-combined(pe).
Undigested pe reads are output but flashed are not written. Defaults to “pe”.</p>
</dd>
<dt class="field-even">type mode</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param output_file</dt>
<dd class="field-odd"><p>Output fastq file path. Defaults to “out.fastq.gz”.</p>
</dd>
<dt class="field-even">type output_file</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param minimum_slice_length</dt>
<dd class="field-odd"><p>Minimum allowed length for in silico digested reads. Defaults to 18.</p>
</dd>
<dt class="field-even">type minimum_slice_length</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param compression_level</dt>
<dd class="field-odd"><p>Compression level for gzip output (1-9). Defaults to 5.</p>
</dd>
<dt class="field-even">type compression_level</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param n_cores</dt>
<dd class="field-odd"><p>Number of digestion processes to use. Defaults to 1.</p>
</dd>
<dt class="field-even">type n_cores</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param read_buffer</dt>
<dd class="field-odd"><p>Number of reads to process before writing to file. Defaults to 100000.</p>
</dd>
<dt class="field-even">type read_buffer</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param stats_prefix</dt>
<dd class="field-odd"><p>Output prefix for stats file. Defaults to “”.</p>
</dd>
<dt class="field-even">type stats_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param keep_cutsite</dt>
<dd class="field-odd"><p>Determines if cutsite is removed from the output. Defaults to False.</p>
</dd>
<dt class="field-even">type keep_cutsite</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param sample_name</dt>
<dd class="field-odd"><p>Name of sample processed eg. DOX-treated_1. Defaults to ‘’.</p>
</dd>
<dt class="field-even">type sample_name</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-digest <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FASTQ<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-r">
<span id="cmdoption-ccanalyser-fastq-digest-restriction_enzyme"></span><span id="cmdoption-ccanalyser-fastq-digest-restriction-enzyme"></span><code class="sig-name descname"><span class="pre">-r</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--restriction_enzyme</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;restriction_enzyme&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-r" title="Permalink to this definition">¶</a></dt>
<dd><p>Restriction enzyme name or sequence</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-m">
<span id="cmdoption-ccanalyser-fastq-digest-mode"></span><code class="sig-name descname"><span class="pre">-m</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--mode</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;mode&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Digestion mode</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>flashed | pe</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-o">
<span id="cmdoption-ccanalyser-fastq-digest-output_file"></span><span id="cmdoption-ccanalyser-fastq-digest-output-file"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_file</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_file&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-o" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-p">
<span id="cmdoption-ccanalyser-fastq-digest-n_cores"></span><span id="cmdoption-ccanalyser-fastq-digest-n-cores"></span><code class="sig-name descname"><span class="pre">-p</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--n_cores</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;n_cores&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-p" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-minimum_slice_length">
<span id="cmdoption-ccanalyser-fastq-digest-minimum-slice-length"></span><code class="sig-name descname"><span class="pre">--minimum_slice_length</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;minimum_slice_length&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-minimum_slice_length" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-keep_cutsite">
<span id="cmdoption-ccanalyser-fastq-digest-keep-cutsite"></span><code class="sig-name descname"><span class="pre">--keep_cutsite</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;keep_cutsite&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-keep_cutsite" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-compression_level">
<span id="cmdoption-ccanalyser-fastq-digest-compression-level"></span><code class="sig-name descname"><span class="pre">--compression_level</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;compression_level&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-compression_level" title="Permalink to this definition">¶</a></dt>
<dd><p>Level of compression for output files</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-read_buffer">
<span id="cmdoption-ccanalyser-fastq-digest-read-buffer"></span><code class="sig-name descname"><span class="pre">--read_buffer</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_buffer&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-read_buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of reads to process before writing to file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-stats_prefix">
<span id="cmdoption-ccanalyser-fastq-digest-stats-prefix"></span><code class="sig-name descname"><span class="pre">--stats_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;stats_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-stats_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for stats file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-sample_name">
<span id="cmdoption-ccanalyser-fastq-digest-sample-name"></span><code class="sig-name descname"><span class="pre">--sample_name</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;sample_name&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-sample_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of sample e.g. DOX_treated_1</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-digest-arg-INPUT_FASTQ">
<span id="cmdoption-ccanalyser-fastq-digest-arg-input-fastq"></span><code class="sig-name descname"><span class="pre">INPUT_FASTQ</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-digest-arg-INPUT_FASTQ" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument(s)</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-fastq-split">
<h3>fastq-split<a class="headerlink" href="#ccanalyser-fastq-split" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Splits fastq file(s) into equal chunks of n reads.</p>
<dl class="field-list simple">
<dt class="field-odd">param input_files</dt>
<dd class="field-odd"><p>Input fastq files to process.</p>
</dd>
<dt class="field-even">type input_files</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param method</dt>
<dd class="field-odd"><p>Use the python or unix (faster but not guarenteed to mantain read pairings between R1 and R2) method to split the fastq files. Defaults to “unix”.</p>
</dd>
<dt class="field-even">type method</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param output_prefix</dt>
<dd class="field-odd"><p>Output prefix for split fastq files. Defaults to “split”.</p>
</dd>
<dt class="field-even">type output_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param compression_level</dt>
<dd class="field-odd"><p>Compression level for gzipped output. Defaults to 5.</p>
</dd>
<dt class="field-even">type compression_level</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param n_reads</dt>
<dd class="field-odd"><p>Number of reads to split the input fastq files into. Defaults to 1000000.</p>
</dd>
<dt class="field-even">type n_reads</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param gzip</dt>
<dd class="field-odd"><p>Gzip compress output files if True. Defaults to True.</p>
</dd>
<dt class="field-even">type gzip</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser fastq-split <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>INPUT_FILES<span class="o">]</span>...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-m">
<span id="cmdoption-ccanalyser-fastq-split-method"></span><code class="sig-name descname"><span class="pre">-m</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--method</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;method&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-m" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to use for splitting</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>python | unix</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-o">
<span id="cmdoption-ccanalyser-fastq-split-output_prefix"></span><span id="cmdoption-ccanalyser-fastq-split-output-prefix"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for deduplicated fastq file(s)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-compression_level">
<span id="cmdoption-ccanalyser-fastq-split-compression-level"></span><code class="sig-name descname"><span class="pre">--compression_level</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;compression_level&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-compression_level" title="Permalink to this definition">¶</a></dt>
<dd><p>Level of compression for output files</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-n">
<span id="cmdoption-ccanalyser-fastq-split-n_reads"></span><span id="cmdoption-ccanalyser-fastq-split-n-reads"></span><code class="sig-name descname"><span class="pre">-n</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--n_reads</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;n_reads&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of reads per fastq file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-gzip">
<span id="cmdoption-ccanalyser-fastq-split-no-gzip"></span><code class="sig-name descname"><span class="pre">--gzip</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--no-gzip</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-gzip" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines if files are gziped or not</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-fastq-split-arg-INPUT_FILES">
<span id="cmdoption-ccanalyser-fastq-split-arg-input-files"></span><code class="sig-name descname"><span class="pre">INPUT_FILES</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-fastq-split-arg-INPUT_FILES" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional argument(s)</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-genome-digest">
<h3>genome-digest<a class="headerlink" href="#ccanalyser-genome-digest" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Generates a bed formatted file of all restriction fragments.</p>
<p>Digests the supplied genome fasta file and generates a bed file containing the locations of all restriction fragments
produced by the supplied restriction enzyme</p>
<dl class="field-list simple">
<dt class="field-odd">param input_fasta</dt>
<dd class="field-odd"><p>Path to fasta file containing whole genome sequence, split by chromosome</p>
</dd>
<dt class="field-even">type input_fasta</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param recognition_site</dt>
<dd class="field-odd"><p>Restriction enzyme name/ Sequence of recognition site.</p>
</dd>
<dt class="field-even">type recognition_site</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param logfile</dt>
<dd class="field-odd"><p>Output path of the digestion logfile. Defaults to genome_digest.log.</p>
</dd>
<dt class="field-even">type logfile</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param output_file</dt>
<dd class="field-odd"><p>Output path for digested chromosome bed file. Defaults to genome_digest.bed.</p>
</dd>
<dt class="field-even">type output_file</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param remove_cutsite</dt>
<dd class="field-odd"><p>Determines if restriction site is removed. Defaults to True.</p>
</dd>
<dt class="field-even">type remove_cutsite</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser genome-digest <span class="o">[</span>OPTIONS<span class="o">]</span> INPUT_FASTA
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-genome-digest-r">
<span id="cmdoption-ccanalyser-genome-digest-recognition_site"></span><span id="cmdoption-ccanalyser-genome-digest-recognition-site"></span><code class="sig-name descname"><span class="pre">-r</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--recognition_site</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;recognition_site&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-genome-digest-r" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Recognition enzyme or sequence</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-genome-digest-l">
<span id="cmdoption-ccanalyser-genome-digest-logfile"></span><code class="sig-name descname"><span class="pre">-l</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--logfile</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;logfile&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-genome-digest-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Path for digestion log file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-genome-digest-o">
<span id="cmdoption-ccanalyser-genome-digest-output_file"></span><span id="cmdoption-ccanalyser-genome-digest-output-file"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_file</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_file&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-genome-digest-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file path</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-genome-digest-remove_cutsite">
<span id="cmdoption-ccanalyser-genome-digest-remove-cutsite"></span><code class="sig-name descname"><span class="pre">--remove_cutsite</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;remove_cutsite&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-genome-digest-remove_cutsite" title="Permalink to this definition">¶</a></dt>
<dd><p>Exclude the recognition sequence from the output</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-genome-digest-arg-INPUT_FASTA">
<span id="cmdoption-ccanalyser-genome-digest-arg-input-fasta"></span><code class="sig-name descname"><span class="pre">INPUT_FASTA</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-genome-digest-arg-INPUT_FASTA" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-interactions-bedgraph">
<h3>interactions-bedgraph<a class="headerlink" href="#ccanalyser-interactions-bedgraph" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Extracts a bedgraph for a specific capture probe/ all capture probes present in a cooler file.</p>
<blockquote>
<div><p>Bedgraph can be normalised for the number of cis interactions and/or binned into even genomic
windows.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">param cooler_fn</dt>
<dd class="field-odd"><p>Path to hdf5 file containing cooler groups.</p>
</dd>
<dt class="field-even">type cooler_fn</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param capture_names</dt>
<dd class="field-odd"><p>Name of capture probe to extract.
If None, will process all probes present in the file.
Defaults to None.</p>
</dd>
<dt class="field-even">type capture_names</dt>
<dd class="field-even"><p>list, optional</p>
</dd>
<dt class="field-odd">param output_prefix</dt>
<dd class="field-odd"><p>Output file prefix for bedgraph. Defaults to “”.</p>
</dd>
<dt class="field-even">type output_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param normalise</dt>
<dd class="field-odd"><p>Normalise counts using the number of cis interactions. Defaults to False.</p>
</dd>
<dt class="field-even">type normalise</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param binsize</dt>
<dd class="field-odd"><p>Genomic binsize to use for generating bedgraph. No binning performed if less than 0. Defaults to 0.</p>
</dd>
<dt class="field-even">type binsize</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param gzip</dt>
<dd class="field-odd"><p>Compress output bedgraph with gzip. Defaults to True.</p>
</dd>
<dt class="field-even">type gzip</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param scale_factor</dt>
<dd class="field-odd"><p>Scaling factor for normalisation. Defaults to 1e6.</p>
</dd>
<dt class="field-even">type scale_factor</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param sparse</dt>
<dd class="field-odd"><p>Produce bedgraph containing just positive bins (True) or all bins (False). Defaults to True.</p>
</dd>
<dt class="field-even">type sparse</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-bedgraph <span class="o">[</span>OPTIONS<span class="o">]</span> COOLER_FN
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-n">
<span id="cmdoption-ccanalyser-interactions-bedgraph-capture_names"></span><span id="cmdoption-ccanalyser-interactions-bedgraph-capture-names"></span><code class="sig-name descname"><span class="pre">-n</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--capture_names</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;capture_names&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Capture to extract and convert to bedgraph, if not provided will transform all.</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-o">
<span id="cmdoption-ccanalyser-interactions-bedgraph-output_prefix"></span><span id="cmdoption-ccanalyser-interactions-bedgraph-output-prefix"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for bedgraphs</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-normalise">
<code class="sig-name descname"><span class="pre">--normalise</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-normalise" title="Permalink to this definition">¶</a></dt>
<dd><p>Normalised bedgraph (Correct for number of cis reads)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-binsize">
<code class="sig-name descname"><span class="pre">--binsize</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-binsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Binsize to use for converting bedgraph to evenly sized genomic bins</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-gzip">
<code class="sig-name descname"><span class="pre">--gzip</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-gzip" title="Permalink to this definition">¶</a></dt>
<dd><p>Compress output using gzip</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-scale_factor">
<span id="cmdoption-ccanalyser-interactions-bedgraph-scale-factor"></span><code class="sig-name descname"><span class="pre">--scale_factor</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;scale_factor&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-scale_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Scale factor to use for bedgraph normalisation</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-sparse">
<span id="cmdoption-ccanalyser-interactions-bedgraph-dense"></span><code class="sig-name descname"><span class="pre">--sparse</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--dense</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-sparse" title="Permalink to this definition">¶</a></dt>
<dd><p>Produce bedgraph containing just positive bins (sparse) or all bins (dense)</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-bedgraph-arg-COOLER_FN">
<span id="cmdoption-ccanalyser-interactions-bedgraph-arg-cooler-fn"></span><code class="sig-name descname"><span class="pre">COOLER_FN</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-bedgraph-arg-COOLER_FN" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-interactions-count">
<h3>interactions-count<a class="headerlink" href="#ccanalyser-interactions-count" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Determines the number of captured restriction fragment interactions genome wide.</p>
<dl class="field-list simple">
<dt class="field-odd">param reporters</dt>
<dd class="field-odd"><p>Reporter tsv file path.</p>
</dd>
<dt class="field-even">type reporters</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output file path for interaction counts tsv. Defaults to ‘counts.tsv’.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param remove_exclusions</dt>
<dd class="field-odd"><p>Removes regions marked as capture proximity exclusions before counting. Defaults to False.</p>
</dd>
<dt class="field-even">type remove_exclusions</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param remove_capture</dt>
<dd class="field-odd"><p>Removes all capture fragments before counting. Defaults to False.</p>
</dd>
<dt class="field-even">type remove_capture</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param subsample</dt>
<dd class="field-odd"><p>Subsamples the fragments by the specified fraction. Defaults to 0 i.e. No subsampling.</p>
</dd>
<dt class="field-even">type subsample</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-count <span class="o">[</span>OPTIONS<span class="o">]</span> REPORTERS
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-count-o">
<span id="cmdoption-ccanalyser-interactions-count-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-count-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of output file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-count-remove_exclusions">
<span id="cmdoption-ccanalyser-interactions-count-remove-exclusions"></span><code class="sig-name descname"><span class="pre">--remove_exclusions</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-count-remove_exclusions" title="Permalink to this definition">¶</a></dt>
<dd><p>Prevents analysis of fragments marked as proximity exclusions</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-count-remove_capture">
<span id="cmdoption-ccanalyser-interactions-count-remove-capture"></span><code class="sig-name descname"><span class="pre">--remove_capture</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-count-remove_capture" title="Permalink to this definition">¶</a></dt>
<dd><p>Prevents analysis of capture fragment interactions</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-count-subsample">
<code class="sig-name descname"><span class="pre">--subsample</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;subsample&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-count-subsample" title="Permalink to this definition">¶</a></dt>
<dd><p>Subsamples reporters before analysis of interactions</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-count-arg-REPORTERS">
<span id="cmdoption-ccanalyser-interactions-count-arg-reporters"></span><code class="sig-name descname"><span class="pre">REPORTERS</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-count-arg-REPORTERS" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-interactions-store">
<h3>interactions-store<a class="headerlink" href="#ccanalyser-interactions-store" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Stores interaction counts.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-store <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<div class="section" id="ccanalyser-interactions-store-bins">
<h4>bins<a class="headerlink" href="#ccanalyser-interactions-store-bins" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Convert cooler file binned at the restriction fragment level to constant genomic windows.</p>
<dl class="field-list simple">
<dt class="field-odd">param cooler_fn</dt>
<dd class="field-odd"><p>Path to cooler file. Nested coolers can be specified by STORE_FN.hdf5::/PATH_TO_COOLER</p>
</dd>
<dt class="field-even">type cooler_fn</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Path for output binned cooler file.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param binsizes</dt>
<dd class="field-odd"><p>Genomic window sizes to use for binning. Defaults to None.</p>
</dd>
<dt class="field-even">type binsizes</dt>
<dd class="field-even"><p>Tuple, optional</p>
</dd>
<dt class="field-odd">param normalise</dt>
<dd class="field-odd"><p>Normalise the number of interactions to total number of cis interactions (True). Defaults to False.</p>
</dd>
<dt class="field-even">type normalise</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
<dt class="field-odd">param n_cores</dt>
<dd class="field-odd"><p>Number of cores to use for binning. Performed in parallel by chromosome. Defaults to 1.</p>
</dd>
<dt class="field-even">type n_cores</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param scale_factor</dt>
<dd class="field-odd"><p>Scaling factor to use for normalising interactions. Defaults to 1e6.</p>
</dd>
<dt class="field-even">type scale_factor</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param overlap_fraction</dt>
<dd class="field-odd"><p>Minimum fraction to use for defining overlapping bins. Defaults to 1e-9.</p>
</dd>
<dt class="field-even">type overlap_fraction</dt>
<dd class="field-even"><p>float, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-store bins <span class="o">[</span>OPTIONS<span class="o">]</span> COOLER_FN
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-b">
<span id="cmdoption-ccanalyser-interactions-store-bins-binsizes"></span><code class="sig-name descname"><span class="pre">-b</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--binsizes</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;binsizes&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Binsizes to use for windowing</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-normalise">
<code class="sig-name descname"><span class="pre">--normalise</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-normalise" title="Permalink to this definition">¶</a></dt>
<dd><p>Enables normalisation of interaction counts during windowing</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-overlap_fraction">
<span id="cmdoption-ccanalyser-interactions-store-bins-overlap-fraction"></span><code class="sig-name descname"><span class="pre">--overlap_fraction</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;overlap_fraction&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-overlap_fraction" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum overlap between genomic bins and restriction fragments for overlap</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-p">
<span id="cmdoption-ccanalyser-interactions-store-bins-n_cores"></span><span id="cmdoption-ccanalyser-interactions-store-bins-n-cores"></span><code class="sig-name descname"><span class="pre">-p</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--n_cores</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;n_cores&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of cores used for binning</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-scale_factor">
<span id="cmdoption-ccanalyser-interactions-store-bins-scale-factor"></span><code class="sig-name descname"><span class="pre">--scale_factor</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;scale_factor&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-scale_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Scaling factor used for normalisation</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-o">
<span id="cmdoption-ccanalyser-interactions-store-bins-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of output file. (Cooler formatted hdf5 file)</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-bins-arg-COOLER_FN">
<span id="cmdoption-ccanalyser-interactions-store-bins-arg-cooler-fn"></span><code class="sig-name descname"><span class="pre">COOLER_FN</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-bins-arg-COOLER_FN" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-interactions-store-fragments">
<h4>fragments<a class="headerlink" href="#ccanalyser-interactions-store-fragments" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Stores restriction fragment interaction combinations at the restriction fragment level.</p>
<dl class="field-list simple">
<dt class="field-odd">param counts</dt>
<dd class="field-odd"><p>Path to restriction fragment interactions counts .tsv file.</p>
</dd>
<dt class="field-even">type counts</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param fragment_map</dt>
<dd class="field-odd"><p>Path to restriction fragment .bed file, generated with genome-digest command.</p>
</dd>
<dt class="field-even">type fragment_map</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output file path for cooler hdf5 file.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param capture_name</dt>
<dd class="field-odd"><p>Name of capture probe.</p>
</dd>
<dt class="field-even">type capture_name</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param capture_oligos</dt>
<dd class="field-odd"><p>Path to capture oligos bed file.</p>
</dd>
<dt class="field-even">type capture_oligos</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param genome</dt>
<dd class="field-odd"><p>Name of genome used for alignment e.g. hg19. Defaults to “”.</p>
</dd>
<dt class="field-even">type genome</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param suffix</dt>
<dd class="field-odd"><p>Suffix to append to filename. Defaults to “”.</p>
</dd>
<dt class="field-even">type suffix</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-store fragments <span class="o">[</span>OPTIONS<span class="o">]</span> COUNTS
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-f">
<span id="cmdoption-ccanalyser-interactions-store-fragments-fragment_map"></span><span id="cmdoption-ccanalyser-interactions-store-fragments-fragment-map"></span><code class="sig-name descname"><span class="pre">-f</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--fragment_map</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;fragment_map&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-f" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Path to digested genome bed file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-c">
<span id="cmdoption-ccanalyser-interactions-store-fragments-capture_oligos"></span><span id="cmdoption-ccanalyser-interactions-store-fragments-capture-oligos"></span><code class="sig-name descname"><span class="pre">-c</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--capture_oligos</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;capture_oligos&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-c" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Path to capture oligos file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-n">
<span id="cmdoption-ccanalyser-interactions-store-fragments-capture_name"></span><span id="cmdoption-ccanalyser-interactions-store-fragments-capture-name"></span><code class="sig-name descname"><span class="pre">-n</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--capture_name</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;capture_name&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-n" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Name of capture oligo to store</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-g">
<span id="cmdoption-ccanalyser-interactions-store-fragments-genome"></span><code class="sig-name descname"><span class="pre">-g</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--genome</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of genome</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-suffix">
<code class="sig-name descname"><span class="pre">--suffix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;suffix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-suffix" title="Permalink to this definition">¶</a></dt>
<dd><p>Suffix to append after the capture name for the output file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-o">
<span id="cmdoption-ccanalyser-interactions-store-fragments-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of output file. (Cooler formatted hdf5 file)</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-fragments-arg-COUNTS">
<span id="cmdoption-ccanalyser-interactions-store-fragments-arg-counts"></span><code class="sig-name descname"><span class="pre">COUNTS</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-fragments-arg-COUNTS" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-interactions-store-merge">
<h4>merge<a class="headerlink" href="#ccanalyser-interactions-store-merge" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Merges ccanalyser cooler files together.</p>
<blockquote>
<div><p>Produces a unified cooler with both restriction fragment and genomic bins whilst
reducing the storage space required by hard linking the “bins” tables to prevent duplication.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">param coolers</dt>
<dd class="field-odd"><p>Cooler files produced by either the fragments or bins subcommands.</p>
</dd>
<dt class="field-even">type coolers</dt>
<dd class="field-even"><p>Tuple</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Path from merged cooler file.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser interactions-store merge <span class="o">[</span>OPTIONS<span class="o">]</span> COOLERS...
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-merge-o">
<span id="cmdoption-ccanalyser-interactions-store-merge-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-merge-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file name</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-interactions-store-merge-arg-COOLERS">
<span id="cmdoption-ccanalyser-interactions-store-merge-arg-coolers"></span><code class="sig-name descname"><span class="pre">COOLERS</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-interactions-store-merge-arg-COOLERS" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument(s)</p>
</dd></dl>

</div>
</div>
<div class="section" id="ccanalyser-reporters-deduplicate">
<h3>reporters-deduplicate<a class="headerlink" href="#ccanalyser-reporters-deduplicate" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Identifies duplicate aligned fragments and removes them.</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser reporters-deduplicate <span class="o">[</span>OPTIONS<span class="o">]</span> COMMAND <span class="o">[</span>ARGS<span class="o">]</span>...
</pre></div>
</div>
<div class="section" id="ccanalyser-reporters-deduplicate-identify">
<h4>identify<a class="headerlink" href="#ccanalyser-reporters-deduplicate-identify" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Identifies read fragments with duplicate coordinates and outputs these to a json file.</p>
<dl class="field-list simple">
<dt class="field-odd">param fragments_fn</dt>
<dd class="field-odd"><p>Input fragments.tsv file to process.</p>
</dd>
<dt class="field-even">type fragments_fn</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output path to output duplicated parental read ids. Defaults to “duplicated_ids.json”.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param buffer</dt>
<dd class="field-odd"><p>Number of fragments to process in memory. Defaults to 1e6.</p>
</dd>
<dt class="field-even">type buffer</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param read_type</dt>
<dd class="field-odd"><p>Process combined(flashed) or non-combined reads (pe).
Due to the low confidence in the quaility of pe reads, duplicates are identified by
removing any fragments with matching start and end coordinates.
Defaults to “flashed”.</p>
</dd>
<dt class="field-even">type read_type</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser reporters-deduplicate identify <span class="o">[</span>OPTIONS<span class="o">]</span> FRAGMENTS_FN
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-identify-o">
<span id="cmdoption-ccanalyser-reporters-deduplicate-identify-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-identify-o" title="Permalink to this definition">¶</a></dt>
<dd><p>File to output fragments with duplicated coordinates</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-identify-buffer">
<code class="sig-name descname"><span class="pre">--buffer</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;buffer&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-identify-buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of fragments to process at one time</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-identify-read_type">
<span id="cmdoption-ccanalyser-reporters-deduplicate-identify-read-type"></span><code class="sig-name descname"><span class="pre">--read_type</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_type&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-identify-read_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of read</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>flashed | pe</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-identify-arg-FRAGMENTS_FN">
<span id="cmdoption-ccanalyser-reporters-deduplicate-identify-arg-fragments-fn"></span><code class="sig-name descname"><span class="pre">FRAGMENTS_FN</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-identify-arg-FRAGMENTS_FN" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-reporters-deduplicate-remove">
<h4>remove<a class="headerlink" href="#ccanalyser-reporters-deduplicate-remove" title="Permalink to this headline">¶</a></h4>
<p class="rubric">Parameters</p>
<p>Removes duplicated fragments from slices file according to duplicated reads found by identify.</p>
<dl class="field-list simple">
<dt class="field-odd">param slices_fn</dt>
<dd class="field-odd"><p>Input slices.tsv file.</p>
</dd>
<dt class="field-even">type slices_fn</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param duplicated_ids</dt>
<dd class="field-odd"><p>Duplicated parental read ids in json format.</p>
</dd>
<dt class="field-even">type duplicated_ids</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output file path for deduplicated slices. Defaults to “dedup.slices.tsv.gz”.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param buffer</dt>
<dd class="field-odd"><p>Number of slices to process in memory. Defaults to 1e6.</p>
</dd>
<dt class="field-even">type buffer</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param sample_name</dt>
<dd class="field-odd"><p>Name of sample being processed e.g. DOX-treated_1 used for statistics. Defaults to “”.</p>
</dd>
<dt class="field-even">type sample_name</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param read_type</dt>
<dd class="field-odd"><p>Process combined(flashed) or non-combined reads (pe) used for statistics. Defaults to “”.</p>
</dd>
<dt class="field-even">type read_type</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param stats_prefix</dt>
<dd class="field-odd"><p>Output path for deduplication statistics. Defaults to “”.</p>
</dd>
<dt class="field-even">type stats_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser reporters-deduplicate remove <span class="o">[</span>OPTIONS<span class="o">]</span> SLICES_FN
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-d">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-duplicated_ids"></span><span id="cmdoption-ccanalyser-reporters-deduplicate-remove-duplicated-ids"></span><code class="sig-name descname"><span class="pre">-d</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--duplicated_ids</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;duplicated_ids&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Path to duplicated ids file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-o">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file for deduplicated slices</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-buffer">
<code class="sig-name descname"><span class="pre">--buffer</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;buffer&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-buffer" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of fragments to process at one time</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-stats_prefix">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-stats-prefix"></span><code class="sig-name descname"><span class="pre">--stats_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;stats_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-stats_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for stats file</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-sample_name">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-sample-name"></span><code class="sig-name descname"><span class="pre">--sample_name</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;sample_name&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-sample_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of sample e.g. DOX_treated_1</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-read_type">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-read-type"></span><code class="sig-name descname"><span class="pre">--read_type</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_type&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-read_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of read</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>flashed | pe</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-deduplicate-remove-arg-SLICES_FN">
<span id="cmdoption-ccanalyser-reporters-deduplicate-remove-arg-slices-fn"></span><code class="sig-name descname"><span class="pre">SLICES_FN</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-deduplicate-remove-arg-SLICES_FN" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
</div>
<div class="section" id="ccanalyser-reporters-identify">
<h3>reporters-identify<a class="headerlink" href="#ccanalyser-reporters-identify" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Removes all non-capture and non-reporter slices to identify reporters.</p>
<dl class="field-list simple">
<dt class="field-odd">param bam</dt>
<dd class="field-odd"><p>Input bam file to analyse</p>
</dd>
<dt class="field-even">type bam</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param annotations</dt>
<dd class="field-odd"><p>Annotations file generated by slices-annotate</p>
</dd>
<dt class="field-even">type annotations</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param output_prefix</dt>
<dd class="field-odd"><p>Output file prefix. Defaults to “reporters”.</p>
</dd>
<dt class="field-even">type output_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param stats_prefix</dt>
<dd class="field-odd"><p>Output stats prefix. Defaults to “”.</p>
</dd>
<dt class="field-even">type stats_prefix</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param method</dt>
<dd class="field-odd"><p>Analysis method. Choose from (capture|tri|tiled). Defaults to “capture”.</p>
</dd>
<dt class="field-even">type method</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param sample_name</dt>
<dd class="field-odd"><p>Sample being processed e.g. DOX-treated_1. Defaults to “”.</p>
</dd>
<dt class="field-even">type sample_name</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param read_type</dt>
<dd class="field-odd"><p>Process combined(flashed) or non-combined reads (pe) used for statistics. Defaults to “”.</p>
</dd>
<dt class="field-even">type read_type</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param gzip</dt>
<dd class="field-odd"><p>Compress output with gzip. Defaults to False.</p>
</dd>
<dt class="field-even">type gzip</dt>
<dd class="field-even"><p>bool, optional</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser reporters-identify <span class="o">[</span>OPTIONS<span class="o">]</span> <span class="o">[</span>capture<span class="p">|</span>tri<span class="p">|</span>tiled<span class="o">]</span>
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-b">
<span id="cmdoption-ccanalyser-reporters-identify-bam"></span><code class="sig-name descname"><span class="pre">-b</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--bam</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;bam&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-b" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Bam file to process</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-a">
<span id="cmdoption-ccanalyser-reporters-identify-annotations"></span><code class="sig-name descname"><span class="pre">-a</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--annotations</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;annotations&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-a" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Required</strong> Annotations for the bam file. Generated by slices-annotate.</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-o">
<span id="cmdoption-ccanalyser-reporters-identify-output_prefix"></span><span id="cmdoption-ccanalyser-reporters-identify-output-prefix"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for deduplicated fastq file(s)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-stats_prefix">
<span id="cmdoption-ccanalyser-reporters-identify-stats-prefix"></span><code class="sig-name descname"><span class="pre">--stats_prefix</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;stats_prefix&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-stats_prefix" title="Permalink to this definition">¶</a></dt>
<dd><p>Output prefix for stats file(s)</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-sample_name">
<span id="cmdoption-ccanalyser-reporters-identify-sample-name"></span><code class="sig-name descname"><span class="pre">--sample_name</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;sample_name&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-sample_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of sample e.g. DOX_treated_1</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-read_type">
<span id="cmdoption-ccanalyser-reporters-identify-read-type"></span><code class="sig-name descname"><span class="pre">--read_type</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;read_type&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-read_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Type of read</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>flashed | pe</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-gzip">
<span id="cmdoption-ccanalyser-reporters-identify-no-gzip"></span><code class="sig-name descname"><span class="pre">--gzip</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--no-gzip</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-gzip" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines if files are gziped or not</p>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-reporters-identify-arg-METHOD">
<span id="cmdoption-ccanalyser-reporters-identify-arg-method"></span><code class="sig-name descname"><span class="pre">METHOD</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-reporters-identify-arg-METHOD" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
<div class="section" id="ccanalyser-slices-annotate">
<h3>slices-annotate<a class="headerlink" href="#ccanalyser-slices-annotate" title="Permalink to this headline">¶</a></h3>
<p class="rubric">Parameters</p>
<p>Annotates a bed file using bedtools intersect.</p>
<p>Enables annotations with both names and counts at the same time
whilst dealing with multimapping reads, which can lead to errors with reporter identification.</p>
<dl class="field-list simple">
<dt class="field-odd">param slices</dt>
<dd class="field-odd"><p>Input bed file.</p>
</dd>
<dt class="field-even">type slices</dt>
<dd class="field-even"><p>os.PathLike</p>
</dd>
<dt class="field-odd">param actions</dt>
<dd class="field-odd"><p>Methods to use for annotation. Choose from (get|count). Defaults to None.</p>
</dd>
<dt class="field-even">type actions</dt>
<dd class="field-even"><p>Tuple, optional</p>
</dd>
<dt class="field-odd">param bed_files</dt>
<dd class="field-odd"><p>Bed files to intersect with the bed file to be annotated. Defaults to None.</p>
</dd>
<dt class="field-even">type bed_files</dt>
<dd class="field-even"><p>Tuple, optional</p>
</dd>
<dt class="field-odd">param names</dt>
<dd class="field-odd"><p>Column names for output tsv file. Defaults to None.</p>
</dd>
<dt class="field-even">type names</dt>
<dd class="field-even"><p>Tuple, optional</p>
</dd>
<dt class="field-odd">param overlap_fractions</dt>
<dd class="field-odd"><p>Minimum overlap fractions required to call an intersection. Defaults to None.</p>
</dd>
<dt class="field-even">type overlap_fractions</dt>
<dd class="field-even"><p>Tuple, optional</p>
</dd>
<dt class="field-odd">param output</dt>
<dd class="field-odd"><p>Output file path for annotated .tsv file. Defaults to None.</p>
</dd>
<dt class="field-even">type output</dt>
<dd class="field-even"><p>os.PathLike, optional</p>
</dd>
<dt class="field-odd">param duplicates</dt>
<dd class="field-odd"><p>Method to deal with multimapping reads/duplicate bed names.
Currently, “remove” is the only supported option. Defaults to “remove”.</p>
</dd>
<dt class="field-even">type duplicates</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">param n_cores</dt>
<dd class="field-odd"><p>Number of corese to use for intersection. Bed files are split by chromosome for faster intersection.
Defaults to 8.</p>
</dd>
<dt class="field-even">type n_cores</dt>
<dd class="field-even"><p>int, optional</p>
</dd>
<dt class="field-odd">param invalid_bed_action</dt>
<dd class="field-odd"><p>Action to deal with invalid bed files. Choose from (ignore|error) .These can be ignored by setting to “ignore”. Defaults to ‘error’.</p>
</dd>
<dt class="field-even">type invalid_bed_action</dt>
<dd class="field-even"><p>str, optional</p>
</dd>
<dt class="field-odd">raises NotImplementedError</dt>
<dd class="field-odd"><p>Only supported option for duplicate bed names is remove.</p>
</dd>
</dl>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>ccanalyser slices-annotate <span class="o">[</span>OPTIONS<span class="o">]</span> SLICES
</pre></div>
</div>
<p class="rubric">Options</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-a">
<span id="cmdoption-ccanalyser-slices-annotate-actions"></span><code class="sig-name descname"><span class="pre">-a</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--actions</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;actions&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Actions to perform for each bed_files file</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>get | count</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-b">
<span id="cmdoption-ccanalyser-slices-annotate-bed_files"></span><span id="cmdoption-ccanalyser-slices-annotate-bed-files"></span><code class="sig-name descname"><span class="pre">-b</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--bed_files</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;bed_files&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Bed files to intersect with slices</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-n">
<span id="cmdoption-ccanalyser-slices-annotate-names"></span><code class="sig-name descname"><span class="pre">-n</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--names</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;names&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Names to use as column names</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-f">
<span id="cmdoption-ccanalyser-slices-annotate-overlap_fractions"></span><span id="cmdoption-ccanalyser-slices-annotate-overlap-fractions"></span><code class="sig-name descname"><span class="pre">-f</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--overlap_fractions</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;overlap_fractions&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-f" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum overlap fractions</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-o">
<span id="cmdoption-ccanalyser-slices-annotate-output"></span><code class="sig-name descname"><span class="pre">-o</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--output</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file name</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-duplicates">
<code class="sig-name descname"><span class="pre">--duplicates</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;duplicates&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to use for reconciling duplicate (i.e. multimapping) slices</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>remove</p>
</dd>
</dl>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-p">
<span id="cmdoption-ccanalyser-slices-annotate-n_cores"></span><span id="cmdoption-ccanalyser-slices-annotate-n-cores"></span><code class="sig-name descname"><span class="pre">-p</span></code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname"><span class="pre">,</span> </code><code class="sig-name descname"><span class="pre">--n_cores</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;n_cores&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-p" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of cores to use for intersections</p>
</dd></dl>

<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-invalid_bed_action">
<span id="cmdoption-ccanalyser-slices-annotate-invalid-bed-action"></span><code class="sig-name descname"><span class="pre">--invalid_bed_action</span></code><code class="sig-prename descclassname"> <span class="pre">&lt;invalid_bed_action&gt;</span></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-invalid_bed_action" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to deal with invalid bed files</p>
<dl class="field-list simple">
<dt class="field-odd">Options</dt>
<dd class="field-odd"><p>ignore | error</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Arguments</p>
<dl class="std option">
<dt id="cmdoption-ccanalyser-slices-annotate-arg-SLICES">
<span id="cmdoption-ccanalyser-slices-annotate-arg-slices"></span><code class="sig-name descname"><span class="pre">SLICES</span></code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-ccanalyser-slices-annotate-arg-SLICES" title="Permalink to this definition">¶</a></dt>
<dd><p>Required argument</p>
</dd></dl>

</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="module_tools/tools.html" class="btn btn-neutral float-right" title="ccanalyser tools" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="running_the_pipeline.html" class="btn btn-neutral float-left" title="Running the pipeline" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, asmith, dsims.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>