

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Pipeline &mdash; ccanalyser 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pipeline FAQ" href="faq_pipeline.html" />
    <link rel="prev" title="Installing" href="installing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> ccanalyser
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installing.html">Installing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-1-create-a-working-directory">Step 1 - Create a working directory</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-edit-a-copy-of-config-yml">Step 2 - Edit a copy of config.yml</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-copy-or-link-fastq-files-into-the-working-directory">Step 3 -  Copy or link fastq files into the <span class="xref std std-term">working directory</span></a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-4-running-the-pipeline">Step 4 - Running the pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-5-running-the-pipeline-to-a-specified-stage">Step 5 - Running the pipeline to a specified stage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="faq_pipeline.html">Pipeline FAQ</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">CLI Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="ccanalyser_module_documentation/modules.html">Module Documentation</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">ccanalyser</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Pipeline</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/running_the_pipeline.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pipeline">
<h1>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this headline">¶</a></h1>
<p>The main feature of ccanalyser is the end-to-end data processing pipeline, the
following diagram illustrates the steps performed by the pipeline:</p>
<a class="reference internal image-reference" href="_images/pipeline_flow.svg"><img alt="Pipeline flow diagram" src="_images/pipeline_flow.svg" width="1000" /></a>
<p>This section provides further details on how to run the pipeline. In essence
the pipeline requires a working directory with correctly named fastq files
and a <a class="reference internal" href="#step-2-edit-a-copy-of-config-yml"><span class="std std-ref">config.yml</span></a> file that provides
the pipeline configuration.</p>
<div class="section" id="step-1-create-a-working-directory">
<h2>Step 1 - Create a working directory<a class="headerlink" href="#step-1-create-a-working-directory" title="Permalink to this headline">¶</a></h2>
<p>To run the pipeline you will need to create a <span class="xref std std-term">working directory</span>
for the pipeline run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">RS411_EPZ5676</span><span class="o">/</span>
<span class="n">cd</span> <span class="n">RS411_EPZ5676</span><span class="o">/</span>
</pre></div>
</div>
<p>The pipeline will be executed here and all files will be generated
in this directory.</p>
</div>
<div class="section" id="step-2-edit-a-copy-of-config-yml">
<h2>Step 2 - Edit a copy of config.yml<a class="headerlink" href="#step-2-edit-a-copy-of-config-yml" title="Permalink to this headline">¶</a></h2>
<p>The configuration file <a class="reference external" href="https://github.com/sims-lab/capture-c/blob/master/config.yml">config.yml</a> enables
parameterisation of the pipeline run with user specific settings. Furthermore,
it also provides paths to essential files for the pipeline run e.g. bowtie2 indicies.
The paths supplied do not have to be in the same directory as the pipeline but it is
recomended to copy the capture oligos used to the <span class="xref std std-term">working directory</span>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The yaml file must be named <strong>config.yml</strong> for the pipeline to recognise it and run correctly.</p>
</div>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="c1">###################################</span>
<span class="c1"># Essential configuration options #</span>
<span class="c1">###################################</span>

<span class="nt">analysis</span><span class="p">:</span>

    <span class="c1"># Method to use for data analysis, choose from capture | tri | tiled</span>
    <span class="nt">method</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">capture</span>

    <span class="c1"># Path to capture oligos used for analysis. This must be in bed format and named.</span>
    <span class="nt">capture_oligos</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">capture-c_oligos.bed</span>

    <span class="c1"># Restriction enzyme name or recognition site</span>
    <span class="nt">restriction_enzyme</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">dpnii</span>

    <span class="c1"># Number of basepairs to exclude around each capture probe to remove re-ligations</span>
    <span class="nt">reporter_exclusion_zone</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">1000</span>

    <span class="c1"># Genomic window size(s) to use for binning restriction fragment interaction counts</span>
    <span class="c1"># into even genomic windows.</span>
    <span class="c1"># Only bin sizes that are present here will be allowed to be used for heatmap generation.</span>
    <span class="nt">bin_size</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2500, 5000</span>

<span class="nt">genome</span><span class="p">:</span>

    <span class="c1"># Name of genome. UCSC genome names are prefered.</span>
    <span class="nt">name</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">mm9</span>

    <span class="c1"># Path to fasta file containing entire genome sequence separated by chromosome.</span>
    <span class="nt">fasta</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/databank/igenomes/Mus_musculus/UCSC/mm9/Sequence/WholeGenomeFasta/genome.fa</span>

    <span class="c1"># Path to indicies for the specified aligner (default = bowtie2)</span>
    <span class="nt">aligner_index</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/databank/igenomes/Mus_musculus/UCSC/mm9/Sequence/Bowtie2Index/genome</span>

    <span class="c1"># Chromosome sizes for genome. Will be determined automatically from the genome name if this is a UCSC genome.</span>
    <span class="nt">chrom_sizes</span><span class="p">:</span>

<span class="c1">#############################</span>
<span class="c1"># Essential cluster options #</span>
<span class="c1">#############################</span>

<span class="nt">pipeline</span><span class="p">:</span>

    <span class="c1"># Cluster manager used. e.g. SLURM, SunGrid Engine, etc.</span>
    <span class="nt">cluster_queue_manager</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">slurm</span>

    <span class="c1"># Name of queue to use. This will be cluster specific.</span>
    <span class="nt">cluster_queue</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">batch</span>

    <span class="c1"># Maximum number of cores to use per task. (~ 4 works best)</span>
    <span class="nt">n_cores</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">4</span>

    <span class="c1"># Maximum memory avalible per job.</span>
    <span class="c1"># Some tasks are quite demanding on memory for a deeply sequenced experiment (32 G recomended)</span>
    <span class="nt">memory</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">32G</span>


<span class="c1">###################################</span>
<span class="c1"># Optional configuration options #</span>
<span class="c1">###################################</span>

<span class="nt">align</span><span class="p">:</span>

    <span class="c1"># Aligner to use. Both bowtie and bowtie2 are supported but bowtie2 is prefered.</span>
    <span class="nt">aligner</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">Bowtie2</span>

    <span class="c1"># Flag to specify index for the aligner. Leave blank if this is not present i.e. bowtie</span>
    <span class="nt">index_flag</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">-x</span>

    <span class="c1"># Aligner specific options. (Take care as this will ignore pipeline_n_cores)</span>
    <span class="nt">options</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">-p 6 --very-sensitive</span>

<span class="nt">analysis_optional</span><span class="p">:</span>

    <span class="c1"># Path to blacklisted regions bed file. Must be named. Can supply any regions to be removed.</span>
    <span class="nt">blacklist</span><span class="p">:</span>

<span class="nt">deduplication</span><span class="p">:</span>

    <span class="c1"># Turns on initial removal of identical reads</span>
    <span class="nt">pre-dedup</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="nt">hub</span><span class="p">:</span>

    <span class="c1"># Determines if hub is created or not.</span>
    <span class="nt">create</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">False</span>

    <span class="c1"># Url/IP of server to host bigWigs</span>
    <span class="nt">url</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">http://userweb.molbiol.ox.ac.uk/</span>

    <span class="c1"># Location of publically accessible location on the server</span>
    <span class="nt">dir</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">/public/asmith/capture-c/test</span>

    <span class="c1"># Name for the hub (UCSC required)</span>
    <span class="nt">name</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">capturec_test</span>

    <span class="c1"># Short hub name (UCSC required)</span>
    <span class="nt">short</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">capturec new pipeline</span>

    <span class="c1"># Long hub name (UCSC required)</span>
    <span class="nt">long</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">capturec processed with the new python pipeline</span>

    <span class="c1"># Email address (UCSC required)</span>
    <span class="nt">email</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">alastair.smith@ndcls.ox.ac.uk</span>

    <span class="c1"># Colours to use for bigWig tracks. Leave blank for random. Colours are cycled if there are more tracks than colours.</span>
    <span class="nt">colors</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">red blue green yellow purple</span>

<span class="nt">normalisation</span><span class="p">:</span>

    <span class="c1"># Scaling factor for normalisation. (not currently used but will be added shortly)</span>
    <span class="nt">scale_factor</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">1000000</span>

<span class="nt">plot</span><span class="p">:</span>

    <span class="c1"># Path to a bed file containing coordinates for plotting a heatmap of reporters.</span>
    <span class="c1"># Must be named and the interval name must contain the probe name to be plotted.</span>
    <span class="nt">coordinates</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">plot_coords.bed</span>

    <span class="c1"># Plot output format (not currently used)</span>
    <span class="nt">format</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">png</span>

    <span class="c1"># Minimum value for plotting</span>
    <span class="nt">vmin</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">0</span>

    <span class="c1"># Maximum value for plotting. Higher values are truncated to this value</span>
    <span class="nt">vmax</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">1</span>

    <span class="c1"># Bin size(s) to use for plotting</span>
    <span class="nt">bin_size</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2500</span>

    <span class="c1"># Normalisation method to use for plot. Leave blank for default based on analysis method</span>
    <span class="nt">normalisation</span><span class="p">:</span>

<span class="nt">trim</span><span class="p">:</span>

    <span class="c1"># Options passed to trim_galore</span>
    <span class="nt">options</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">--length 21</span>

<span class="nt">split</span><span class="p">:</span>

    <span class="c1"># Fastq files are split for parallel processing. Defines number of reads per fastq file (lower = more files to process)</span>
    <span class="nt">n_reads</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">2500000</span>
</pre></div>
</div>
<p>This yaml file can be edited using standard text editors e.g.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># To use gedit</span>
<span class="n">gedit</span> <span class="n">config</span><span class="o">.</span><span class="n">yml</span>

<span class="c1"># To use nano</span>
<span class="n">nano</span> <span class="n">config</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
</div>
<div class="section" id="step-3-copy-or-link-fastq-files-into-the-working-directory">
<h2>Step 3 -  Copy or link fastq files into the <span class="xref std std-term">working directory</span><a class="headerlink" href="#step-3-copy-or-link-fastq-files-into-the-working-directory" title="Permalink to this headline">¶</a></h2>
<p>The pipeline requires that fastq files are paired and in any of these formats:</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Gziped files are handled appropriately without the need for extraction if .gz is
present at the end of the file name.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Multi-lane fastq files should be
concatenated prior to running the pipeline otherwise multiple separate analyses will
be performed.</p>
</div>
<ul class="simple">
<li><p>samplename_R1.fastq.gz</p></li>
<li><p>samplename_1.fastq.gz</p></li>
<li><p>samplename_R1.fastq</p></li>
<li><p>samplename_1.fastq</p></li>
</ul>
<p>All fastq files present in the directory will be processed by the pipeline in parallel and
original fastq files will not be modified. If new fastq files are added to a pre-run pipeline,
only the new files will be processed.</p>
<p>Copy:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cp</span> <span class="n">PATH_TO_FASTQ</span><span class="o">/</span><span class="n">example_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span>
</pre></div>
</div>
<p>Symlink example:</p>
<p>Be sure to use the absolute path for symlinks</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="n">ABSOLUTE_PATH_TO_FASTQ</span><span class="o">/</span><span class="n">example_R1</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</div>
<div class="section" id="step-4-running-the-pipeline">
<h2>Step 4 - Running the pipeline<a class="headerlink" href="#step-4-running-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>After copying/linking fastq files into the working directory and configuring
config.yml for the current experiment, the pipeline can be run with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ccanalyser</span> <span class="n">pipeline</span>
</pre></div>
</div>
<p>There are several options to visualise which tasks will be performed by the pipeline
before running.</p>
<p>The tasks to be performed can be examined with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Shows the tasks to be performed</span>
<span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">show</span>

<span class="c1"># Plots a directed graph using graphviz</span>
<span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">plot</span>
</pre></div>
</div>
<p>If you are happy with the tasks to be performed, the full pipeline run can be started with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># If using all default settings and using a cluster</span>
<span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">make</span>

<span class="c1"># If not using a cluster, run in local mode.</span>
<span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">make</span> <span class="o">--</span><span class="n">local</span> <span class="o">-</span><span class="n">p</span> <span class="mi">4</span>

<span class="c1"># Avoiding disconnects</span>
<span class="n">nohup</span> <span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">make</span> <span class="o">&amp;</span>
</pre></div>
</div>
<p>See <a class="reference external" href="https://cgat-core.readthedocs.io/en/latest/getting_started/Examples.html">cgatcore</a> for additional
information.</p>
</div>
<div class="section" id="step-5-running-the-pipeline-to-a-specified-stage">
<h2>Step 5 - Running the pipeline to a specified stage<a class="headerlink" href="#step-5-running-the-pipeline-to-a-specified-stage" title="Permalink to this headline">¶</a></h2>
<p>There are currently multiple stopping points built into the pipeline at key stages. These are:</p>
<ul class="simple">
<li><p>fastq_preprocessing - Stops after in silico digestion of fastq files.</p></li>
<li><p>pre_annotation - Stops before aligned slices are ready to be annotated.</p></li>
<li><p>post_annotation - Stops after aligned slices have been annotated.</p></li>
<li><p>post_ccanalyser_analysis - Stops after reporters have been identified and duplicate filtered.</p></li>
<li><p>full - Run the pipline until all required tasks are complete</p></li>
</ul>
<p>To run the pipeline until one of these stopping points, use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run until TASK_NAME step</span>
<span class="n">ccanalyser</span> <span class="n">pipeline</span> <span class="n">make</span> <span class="n">TASK_NAME</span>
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="faq_pipeline.html" class="btn btn-neutral float-right" title="Pipeline FAQ" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="installing.html" class="btn btn-neutral float-left" title="Installing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2020, asmith, dsims.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>